[FEATURE] New Patrol Tags: all_mentored and app{index}_mentored

CL0WNTH0UGHTS opened the pull request:
<!-- Write BELOW The Headers and ABOVE The comments else it may not be viewable. -->
<!-- You can view CONTRIBUTING.md for a detailed description of the pull request process. -->
<!-- Be sure to name your PR something descriptive and succinct; include Bugfix: Feature: Enhancement: or Content: in the title to describe what type of PR it is. -->
<!-- IF YOU ARE DOING A BUGFIX: Please target the latest release branch if the bug that you are fixing is also present in the latest release. -->

## About The Pull Request
Introduces a system of new tags, catered towards patrols, which check the status of the patrol apprentices mentor status:
**"all_mentored"** - checks if every single apprentice within the patrols cats has a mentor. If even a single one doesnt, patrols tagged with this will be removed from the possible pool of patrols
**"app{index}_mentored"** - checks if the apprentice at the specific index slot (such as if you have five apprentices on a patrol, and one of the chosen patrols are tagged with "app1_mentored", it will check the very first apprentice within the patrol cats list): if the apprentice at the index spot is not mentored/has no mentor, the patrol is removed from the possible pool of patrols
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage senior developers from merging your PR! -->

## Why This Is Good For ClanGen
The new tags help prevent possible bugs/awkward texts on the players side (where an unmentored cat mentions a mentor) and also allows contributors to write more specific scenarios - ala a patrol where a warrior apprentice and a medicine cat apprentice exchange gossip/tips/other such facts about their own given mentor (like how students do in school about their teachers). 

<!-- If this is a bug fix, you can remove this section. -->
<!-- Please add a short description of why you think these changes would benefit the game. -->
<!-- If this PR was the result of discussion/testing on the discord, please add a link to the discord conversation here. -->
<!-- If you have multiple features that can stand on their own, or unrelated bugfixes, please create separate PRs for them. -->

## Linked Issues
#3932
#3953
<!-- If this is unrelated to an issue, you can remove this section. -->
<!-- If this was in response to a GitHub issue, please write it here with the format Fixes: #1234 so that GitHub knows to link the issues. -->
<!-- If this PR was the result of discussion/testing on the discord, please add a link to the discord conversation here. -->

## Proof of Testing
https://screenrec.com/share/iAjgxt37r6
**DEBUG OUTPUT AS PROOF:**
PATROL START ---------------------------------------------------
Patrol Leader: Daisypaw
Random Cat: Daisypaw
debug_ensure_patrol_id: "mtn_med_appfirstsign1" is a possible herb_gathering patrol, and was set as the only herb_gathering patrol option
Total Number of Possible Patrols | normal: 1, romantic: 1
not a potential mate or current mate
starting chance: 50 | EX_updated chance: 55
Daisypaw updated chance to 55 |
PATROL ID: mtn_med_appfirstsign1 | SUCCESS: False
PATROL END -----------------------------------------------------
PATROL START ---------------------------------------------------
Patrol Leader: Blackpaw
Random Cat: Blackpaw
DEBUG: requested patrol does not meet constraints (tags)
debug_ensure_patrol_id: "mtn_med_appfirstsign1" is not found. Check output for reason.

<!-- Include any screenshots, debugging steps, or links to beta testing threads here. At least one form of proof of testing is REQUIRED for all new content. You must be able to run the code locally before you PR it here. -->

## Changelog/Credits
- Including two (technically 7) new patrol tags: all_mentored and app{index}_mentored. 
<!-- Include any changes that should be made to the changelog of the game here or any changes to the credits file of the game. -->
<!-- This is just for easy access later for senior developers gathering this information; this process is not automated. -->

CL0WNTH0UGHTS committed:
Create A 'has_mentor' Check

Creates two different tags: "app_mentored" (which checks if all apprentices thats within the patrol_apprentices has a mentor) and "app{index}_mentored" (which checks if the apprentice at a specific number within the app list has a mentor).

CL0WNTH0UGHTS committed:
Update (Patrol) Tag Filtering

Updates event_filtering to check if the apprentices have a mentor

CL0WNTH0UGHTS committed:
Update Docs & Schema

Updates patrol document to include the desc. of both tags, and update the patrol schema with all the new patrol tags.

CL0WNTH0UGHTS committed:
Update (Most) Beach Patrols

If any patrol mentioned mentor(s) which indicated ownership from cat(s) in the patrol, it was tagged.

CL0WNTH0UGHTS committed:
Tag (Most) Desert Patrols

If any patrol mentioned mentor(s) which indicated ownership from cat(s) in the patrol, it was tagged.

CL0WNTH0UGHTS committed:
Tag (Most) Forest Patrols

If any patrol mentioned mentor(s) which indicated ownership from cat(s) in the patrol, it was tagged.

CL0WNTH0UGHTS committed:
Tag (Most) General Patrols

If any patrol mentioned mentor(s) which indicated ownership from cat(s) in the patrol, it was tagged.

CL0WNTH0UGHTS committed:
Tag (Most) Mountainous Patrols

If any patrol mentioned mentor(s) which indicated ownership from cat(s) in the patrol, it was tagged.

CL0WNTH0UGHTS committed:
Tag (Most) Plains Patrols

If any patrol mentioned mentor(s) which indicated ownership from cat(s) in the patrol, it was tagged.

CL0WNTH0UGHTS committed:
Tag (Most) Wetlands Patrols

If any patrol mentioned mentor(s) which indicated ownership from cat(s) in the patrol, it was tagged.

CL0WNTH0UGHTS committed:
Fix Tests Failing

Fixes some formatting errors that were causing a unit test to fail

scribblecrumb commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -584,6 +584,18 @@ def _filter_patrols(
+        if self.patrol_apprentices:
+            general_mentor_checks = all(app.mentor for app in self.patrol_apprentices)
+        else:
+            general_mentor_checks = False
```
This could be shortened down to just:
```python
general_mentor_checks = all(app.mentor for app in self.patrol_apprentices) if self.patrol_apprentices else False
```

CL0WNTH0UGHTS replied:
Thank you for the note! I've applied it :D

scribblecrumb commented on scripts/events_module/event_filters.py:
```diff
@@ -136,6 +136,13 @@ def event_for_tags(tags: list, cat, other_cat=None) -> bool:
+    for _tag in tags:
+        is_app_mentored = re.match(r"app[1-6]_mentored", _tag)
+        if "all_mentored" in tags and not has_mentor.get("general", False):
+            return False
```
This could be shortened to:
```python
        if "all_mentored" in tags:
            return has_mentor.get("general", False)
```

CL0WNTH0UGHTS replied:
I believe I applied what you requested (Which, thank you for). But, please tell me if I still have some part of this to clean up :).

CL0WNTH0UGHTS committed:
Fix/Apply Requested Changes

Apply the changes requested by reviewers - makes the code more streamlined and simplified.

CL0WNTH0UGHTS committed:
Fix Test Fail

Once again, fix my formatting in order for the test to run.

scribblecrumb commented on scripts/events_module/event_filters.py:
```diff
@@ -50,7 +50,7 @@ def event_for_season(seasons: list) -> bool:
 
 
-def event_for_tags(tags: list, cat, other_cat=None) -> bool:
+def event_for_tags(tags: list, cat, other_cat=None, has_mentor=None) -> bool:
```
let's change the `has_mentor` param to `mentor_tags_fulfilled` for readability, cus I keep getting lost about what it means and having to check lol

CL0WNTH0UGHTS replied:
Hopefully implemented it properly! But, I agree with that - I worried that the has_mentor was too nebulous especially when disconnected from the original check.

CL0WNTH0UGHTS committed:
Update With Requested Changes

Update the has_mentor tag in the function event_for_tags to mentor_tags_fulfilled.

CL0WNTH0UGHTS committed:
Fix a Missed has_mentor Call

missed a has_mentor whilst rushing to fix it :3
